File: android/app/src/main/java/com/tabletopforge/data/Resources.kt
// Class resource counters — Bardic Inspiration, Ki, Sorcery Points,
// Channel Divinity, Action Surge, Lay on Hands, Wild Shape, etc.
//
// Each definition computes its `total` from the character's level on
// that class. The "used" count lives on the device (DataStore-backed)
// keyed by creatureId + resource id, so the player can tap +/- without
// a round-trip; the GM doesn't typically need a live readout of these.
//
// Resets follow the SRD's short/long-rest rules (the rest_kind field).
// Reset buttons aren't auto-fired — players hit them after the table
// agrees a rest happened.

package com.tabletopforge.data

enum class RestKind { SHORT, LONG }

data class ResourceDef(
    val id: String,
    val label: String,
    val total: (level: Int) -> Int,
    val rest: RestKind,
    val note: String? = null,
)

private fun cha(c: Creature) = ((c.charisma     ?: 10) - 10) / 2
private fun wis(c: Creature) = ((c.wisdom       ?: 10) - 10) / 2
private fun con(c: Creature) = ((c.constitution ?: 10) - 10) / 2

// Map char_class (case-sensitive — matches the SRD names from
// utils/classes.js BASE_CLASSES) to its resource list. Custom classes
// from the custom-classes plugin won't appear here yet — they'd need a
// JSON definition uploaded alongside the choices, which is a follow-up.
fun resourcesFor(c: Creature): List<Pair<ResourceDef, Int>> {
    val cls   = c.char_class?.trim() ?: return emptyList()
    val level = (c.char_level ?: 1).coerceAtLeast(1)
    val defs: List<ResourceDef> = when (cls) {
        "Barbarian" -> listOf(
            ResourceDef("rages", "Rages",
                total = { lvl -> when {
                    lvl >= 17 -> 6
                    lvl >= 12 -> 5
                    lvl >= 6  -> 4
                    lvl >= 3  -> 3
                    else      -> 2
                } },
                rest = RestKind.LONG,
            ),
        )
        "Bard" -> listOf(
            ResourceDef("bardic-inspiration", "Bardic Inspiration",
                total = { _ -> (cha(c)).coerceAtLeast(1) },
                rest = RestKind.SHORT,
                note = "Recharges on a short or long rest.",
            ),
        )
        "Cleric" -> listOf(
            ResourceDef("channel-divinity", "Channel Divinity",
                total = { lvl -> if (lvl >= 18) 3 else if (lvl >= 6) 2 else 1 },
                rest = RestKind.SHORT,
            ),
        )
        "Druid" -> listOf(
            ResourceDef("wild-shape", "Wild Shape",
                total = { lvl -> if (lvl >= 20) Int.MAX_VALUE else 2 },
                rest = RestKind.SHORT,
                note = "2/short rest until level 20 (then unlimited).",
            ),
            ResourceDef("channel-nature", "Channel Nature",
                total = { lvl -> if (lvl >= 18) 3 else if (lvl >= 6) 2 else 1 },
                rest = RestKind.SHORT,
            ),
        )
        "Fighter" -> listOf(
            ResourceDef("action-surge", "Action Surge",
                total = { lvl -> if (lvl >= 17) 2 else 1 },
                rest = RestKind.SHORT,
            ),
            ResourceDef("second-wind", "Second Wind",
                total = { lvl -> if (lvl >= 17) 4 else if (lvl >= 10) 3 else 2 },
                rest = RestKind.SHORT,
            ),
            ResourceDef("indomitable", "Indomitable",
                total = { lvl -> if (lvl >= 17) 3 else if (lvl >= 13) 2 else 1 },
                rest = RestKind.LONG,
                note = "Available from level 9.",
            ),
        )
        "Monk" -> listOf(
            ResourceDef("focus-points", "Focus Points",
                total = { lvl -> lvl.coerceAtLeast(2).let { if (it >= 2) it else 0 } },
                rest = RestKind.SHORT,
                note = "Equal to your monk level (2nd level+).",
            ),
        )
        "Paladin" -> listOf(
            ResourceDef("lay-on-hands", "Lay on Hands (HP pool)",
                total = { lvl -> 5 * lvl },
                rest = RestKind.LONG,
            ),
            ResourceDef("channel-divinity", "Channel Divinity",
                total = { lvl -> if (lvl >= 11) 3 else if (lvl >= 7) 2 else 1 },
                rest = RestKind.SHORT,
                note = "Available from level 3.",
            ),
        )
        "Ranger" -> listOf(
            ResourceDef("hunters-mark-uses", "Hunter's Mark uses",
                total = { _ -> wis(c).coerceAtLeast(1) },
                rest = RestKind.LONG,
                note = "Free casts equal to your WIS modifier.",
            ),
        )
        "Rogue" -> emptyList()
        "Sorcerer" -> listOf(
            ResourceDef("sorcery-points", "Sorcery Points",
                total = { lvl -> if (lvl >= 2) lvl else 0 },
                rest = RestKind.LONG,
            ),
        )
        "Warlock" -> listOf(
            ResourceDef("genies-vessel", "Genie's Vessel charges",
                total = { _ -> 1 },
                rest = RestKind.LONG,
                note = "Replace with whichever Patron-specific feature applies.",
            ),
        )
        "Wizard" -> listOf(
            ResourceDef("arcane-recovery", "Arcane Recovery",
                total = { _ -> 1 },
                rest = RestKind.LONG,
                note = "Once per long rest, on a short rest.",
            ),
        )
        "Artificer" -> listOf(
            ResourceDef("infusions-known",  "Infusions known",
                total = { lvl -> when {
                    lvl >= 18 -> 6; lvl >= 14 -> 5; lvl >= 10 -> 4; lvl >= 6 -> 3; else -> 2
                } },
                rest = RestKind.LONG,
            ),
        )
        else -> emptyList()
    }
    return defs.map { def -> def to def.total(level) }
}

File: frontend/src/data/resources.js
// Class resource counters — Bardic Inspiration, Ki, Sorcery Points,
// Channel Divinity, Action Surge, Lay on Hands, Wild Shape, etc.
//
// Each definition computes its `total` from the character's level and
// stat block. The mutable `used` count lives on the creature row in
// `resource_state` JSONB so it round-trips through the same save path
// as everything else and stays in sync between web and mobile.
//
// Mirror of the Android catalogue at
// android/app/src/main/java/com/tabletopforge/data/Resources.kt —
// keep them aligned when fields move.

export const REST_KIND = {
  SHORT: 'short',
  LONG:  'long',
};

const cha = (c) => Math.floor(((c?.charisma     ?? 10) - 10) / 2);
const wis = (c) => Math.floor(((c?.wisdom       ?? 10) - 10) / 2);

// Bardic Inspiration die size scales with bard level (d6 → d8 → d10 → d12).
function biDie(level) {
  if (level >= 15) return 'd12';
  if (level >= 10) return 'd10';
  if (level >= 5)  return 'd8';
  return 'd6';
}

// Returns `[{ def, total }]` for the character. Custom-classes plugin
// classes don't appear here (they'd need their own resource definitions
// uploaded — follow-up).
export function resourcesForCreature(creature) {
  const cls = (creature?.char_class || '').trim();
  if (!cls) return [];
  const level = Math.max(1, Number(creature?.char_level) || 1);

  const defs = [];
  switch (cls) {
    case 'Barbarian':
      defs.push({
        id: 'rages',
        label: 'Rages',
        total: (lvl => lvl >= 17 ? 6 : lvl >= 12 ? 5 : lvl >= 6 ? 4 : lvl >= 3 ? 3 : 2)(level),
        rest: REST_KIND.LONG,
      });
      break;

    case 'Bard':
      defs.push({
        id: 'bardic-inspiration',
        label: `Bardic Inspiration (${biDie(level)})`,
        // Exposed separately so the "Grant" UI can PATCH the
        // recipient's `inspiration_die` field with the right size.
        die: biDie(level),
        total: Math.max(1, cha(creature)),
        // Long rest until Font of Inspiration kicks in at level 5.
        rest: level >= 5 ? REST_KIND.SHORT : REST_KIND.LONG,
        note: level < 5
          ? 'Recharges on a long rest. (Short rest from level 5.)'
          : 'Recharges on a short or long rest.',
      });
      break;

    case 'Cleric':
      defs.push({
        id: 'channel-divinity',
        label: 'Channel Divinity',
        total: level >= 18 ? 3 : level >= 6 ? 2 : 1,
        rest: REST_KIND.SHORT,
      });
      break;

    case 'Druid':
      defs.push({
        id: 'wild-shape',
        label: 'Wild Shape',
        total: level >= 20 ? Infinity : 2,
        rest: REST_KIND.SHORT,
        note: '2/short rest until level 20 (then unlimited).',
      });
      defs.push({
        id: 'channel-nature',
        label: 'Channel Nature',
        total: level >= 18 ? 3 : level >= 6 ? 2 : 1,
        rest: REST_KIND.SHORT,
      });
      break;

    case 'Fighter':
      defs.push({
        id: 'action-surge',
        label: 'Action Surge',
        total: level >= 17 ? 2 : 1,
        rest: REST_KIND.SHORT,
      });
      defs.push({
        id: 'second-wind',
        label: 'Second Wind',
        total: level >= 17 ? 4 : level >= 10 ? 3 : 2,
        rest: REST_KIND.SHORT,
      });
      defs.push({
        id: 'indomitable',
        label: 'Indomitable',
        total: level >= 17 ? 3 : level >= 13 ? 2 : 1,
        rest: REST_KIND.LONG,
        note: 'Available from level 9.',
      });
      break;

    case 'Monk':
      defs.push({
        id: 'focus-points',
        label: 'Focus Points',
        total: level >= 2 ? level : 0,
        rest: REST_KIND.SHORT,
        note: 'Equal to your monk level (2nd level+).',
      });
      break;

    case 'Paladin':
      defs.push({
        id: 'lay-on-hands',
        label: 'Lay on Hands (HP pool)',
        total: 5 * level,
        rest: REST_KIND.LONG,
      });
      defs.push({
        id: 'channel-divinity',
        label: 'Channel Divinity',
        total: level >= 11 ? 3 : level >= 7 ? 2 : 1,
        rest: REST_KIND.SHORT,
        note: 'Available from level 3.',
      });
      break;

    case 'Ranger':
      defs.push({
        id: 'hunters-mark-uses',
        label: "Hunter's Mark uses",
        total: Math.max(1, wis(creature)),
        rest: REST_KIND.LONG,
        note: 'Free casts equal to your WIS modifier.',
      });
      break;

    case 'Sorcerer':
      defs.push({
        id: 'sorcery-points',
        label: 'Sorcery Points',
        total: level >= 2 ? level : 0,
        rest: REST_KIND.LONG,
      });
      break;

    case 'Wizard':
      defs.push({
        id: 'arcane-recovery',
        label: 'Arcane Recovery',
        total: 1,
        rest: REST_KIND.LONG,
        note: 'Once per long rest, on a short rest.',
      });
      break;

    case 'Artificer':
      defs.push({
        id: 'infusions-known',
        label: 'Infusions known',
        total: level >= 18 ? 6 : level >= 14 ? 5 : level >= 10 ? 4 : level >= 6 ? 3 : 2,
        rest: REST_KIND.LONG,
      });
      break;
  }

  return defs.map((def) => ({ def, total: def.total }));
}
